aws/s3.go

```diff
@@ -88,55 +97,131 @@ func getAllS3Buckets(awsSession *session.Session, excludeAfter time.Time, bucket
 	var bucketNamesPerRegion = make(map[string][]*string)
 	var regionClients = make(map[string]*s3.S3)
 
-	for _, bucket := range output.Buckets {
-		logging.Logger.Debugf("Checking - Bucket %s", *bucket.Name)
+	for _, region := range targetRegions {
+		logging.Logger.Debugf("S3 - creating session - region %s", region)
+		awsSession, err := session.NewSession(&awsgo.Config{
+			Region: awsgo.String(region)},
+		)
+		if err != nil {
+			return bucketNamesPerRegion, errors.WithStackTrace(err)
+		}
+		regionClients[region] = s3.New(awsSession)
+	}
+
+	totalBuckets := len(output.Buckets)
+	if totalBuckets == 0 {
+		return bucketNamesPerRegion, nil
+	}
+
+	if batchSize <= 0 {
+		return nil, fmt.Errorf("Invalid batchsize - %d - should be > 0", batchSize)
+	}
+
+	totalBatches := totalBuckets / batchSize
+	if totalBuckets%batchSize != 0 {
+		totalBatches = totalBatches + 1
+	}
+	batchCount := 1
+
+	// Batch the get operation
+	for i := 0; i < totalBuckets; i += batchSize {
+		j := i + batchSize
+		if j > totalBuckets {
+			j = totalBuckets
+		}
+
+		logging.Logger.Infof("Getting - %d-%d buckets of batch %d/%d", i+1, j, batchCount, totalBatches)
+
+		targetBuckets := output.Buckets[i:j]
+
+		var bucketCh = make(chan *S3Bucket, len(targetBuckets))
+		var wg sync.WaitGroup
 
-		if len(bucketNameSubStr) > 0 {
-			if !strings.Contains(*bucket.Name, bucketNameSubStr) {
+		for _, bucket := range targetBuckets {
+			if len(bucketNameSubStr) > 0 && !strings.Contains(*bucket.Name, bucketNameSubStr) {
 				logging.Logger.Debugf("Skipping - Bucket %s - failed substring filter - %s", *bucket.Name, bucketNameSubStr)
 				continue
 			}
-		}
 
-		if !excludeAfter.After(*bucket.CreationDate) {
-			logging.Logger.Debugf("Skipping - Bucket %s - matched CreationDate filter", *bucket.Name)
-			continue
+			wg.Add(1)
+			go func(bucket *s3.Bucket) {
+				defer wg.Done()
+				if err := getBucketInfo(svc, bucket, regionClients, bucketCh); err != nil {
+					logging.Logger.Warnf("Skipping - Bucket %s - Failed to get info - %s", *bucket.Name, err)
+				}
+			}(bucket)
 		}
 
-		bucketRegion, err := getS3BucketRegion(svc, bucket.Name)
-		if err != nil {
-			logging.Logger.Warnf("Skipping - Bucket %s - Failed to get bucket location.", *bucket.Name)
-			continue
-		}
+		go func() {
+			wg.Wait()
+			close(bucketCh)
+		}()
 
-		// Create session for current bucket region as bucket location's session is required for getting bucket tags.
-		if _, ok := regionClients[bucketRegion]; !ok {
-			logging.Logger.Debugf("Creating session - Bucket %s - region %s", *bucket.Name, bucketRegion)
-			awsSession, err := session.NewSession(&awsgo.Config{
-				Region: awsgo.String(bucketRegion)},
-			)
-			if err != nil {
-				return bucketNamesPerRegion, errors.WithStackTrace(err)
+		for b := range bucketCh {
```

Comment:
This function has gotten really, really long. Consider breaking into separate pieces: e.g., one that fetches the data concurrently; one that processes it; one that orchestrates the whole thing.